### DISCUSSION TITLE
How can I dynamically set environment variables?

### DISCUSSION THREAD

**Discussion by fdw** [creativeprojects/resticprofile#305] [Q&A]
I'm using AWS SSO, which isn't natively supported by `restic`. Therefore, I need to set some `AWS_` env variables manually, before the actual `restic` command is executed.

I've tried to do that in a `run-before`, but then the variables only seem to be set in a subshell and vanish again when that is done. 

Is there some other way to run a script which sets environment variables for the `restic` command?

**Comment by creativeprojects** [comment]
Hi,

Have you tried using the `AWS_PROFILE` environment variable? This is [supported by restic](https://restic.readthedocs.io/en/stable/040_backup.html#environment-variables).

Basically, you need to create a profile in `~/.aws/config` like:

```
[profile sandbox]
sso_start_url = https://aaaaaaaaa.awsapps.com/start
sso_region = us-west-2
sso_account_id = 000000000001
sso_role_name = Admin
region = us-west-2
```

This profile makes use of SSO but you just need to configure it with your authentication mechanism.

From there you can login to AWS from the terminal:

```
aws --profile=sandbox sso login
```

You need to set your `AWS_PROFILE` environment variable to the profile name, either directly in your shell environment:

```
export AWS_PROFILE=sandbox
```

... or in the resticprofile configuration:

```yaml
profile_to_aws:
  env:
    AWS_PROFILE: sandbox
```

The temporary credentials will be stored in an sso cache file which will be picked up by restic.

If you don't use SSO I think you can still store your temporary credentials in the `~/.aws/credentials` file as well.

---

Now, about your original question:
this is something that has been asked a few times and we'll look into this 👍🏻

**Reply by fdw** [reply]
Thank you! 🙂

I've tried it with `AWS_PROFILE`, but that also doesn't seem to work, unfortunately. As far as I can see, `restic` is not able to read the credentials from the SSO cache file (only from `.aws/credentials`, which is not for SSO). But I'd be happy if I was wrong here.

**Reply by creativeprojects** [reply]
Yeah you're right. restic is not using the AWS sdk. The sso cache is not supported by the minio sdk used in restic 😢 

Back to the drawing board... 🤔

**Reply by creativeprojects** [reply]
https://github.com/creativeprojects/resticprofile/issues/316: That would solve your problem, wouldn't it?

**Reply by fdw** [reply]
#316 would help, yes. Another option would be to define a command for an environment variable.

Also, for my specific problem I've opened https://github.com/restic/restic/issues/4688 :) But I still think a dynamic way to send environment variables could be helpful.

**Reply by jkellerer** [reply]
@fdw , I've updated #316 with the intent to support declaring env variables from `run-before`. The syntax will be something like:

```yaml
default:
  run-before: 'echo "VAR_NAME=VALUE" >> {{ env }}'
```

... where `{{ env }}` is a temporary file in ".env" format that is created and removed by resticprofile with perms 0600

**Reply by fdw** [reply]
Interesting concept, and would probably work for me :)
